codeU3.py

```diff
@@ -0,0 +1,222 @@
+        Returns:
+            True: if route gives string that is a prefix in our dictionary
+            False: otherwise
+        '''
+
+        if not route:
+            return False
+        chars = [self.grid[x, y] for (x, y) in route]
+        prefix = ''.join(chars)
+        if self.dict.isPrefix(prefix):
+            return True
+        return False
+
+    def route_word(self, route):
+
+        '''Given a list of tuple coordinates (x,y) encoding route
+        Returns:
+            empty string, if corresponding string is not in the dictionary
+            word, otherwise
+        '''
+
+        if not route:
+            return ""
+        chars = [self.grid[x, y] for (x, y) in route]
+        word = ''.join(chars)
+        if self.dict.isWord(word):
+            return word
+        return ""
+
+    def route_extend(self, route):
+
+        '''Given a route as list of (x,y) indices
+        Returns:
+            all possible routes where 1 step made in one of 8 directions
+            each route is a list of tuple coordinates 
+        '''
+
+        last_x = route[-1][0] #last route coordinate x
+        last_y = route[-1][1] #last route coordinate y
+        cells_set = set(route)
+        route_extended = []
+
+        if last_x + 1 < self.m and (last_x + 1, last_y) not in cells_set: #ensure we don't visit the same cell twice
+            route_extended.append(route + [(last_x + 1, last_y)])
+        if last_x + 1 < self.m and last_y + 1 < self.n and (last_x + 1, last_y + 1) not in cells_set:
+            route_extended.append(route + [(last_x + 1, last_y + 1)])
+        if last_y + 1 < self.n and (last_x, last_y + 1) not in cells_set:
+            route_extended.append(route + [(last_x, last_y + 1)])
+        if last_x - 1 >= 0 and (last_x - 1, last_y) not in cells_set:
+            route_extended.append(route + [(last_x - 1, last_y)])
+        if last_x - 1 >= 0 and last_y + 1 < self.n and (last_x - 1, last_y + 1) not in cells_set:
+            route_extended.append(route+[(last_x - 1, last_y + 1)])
+        if last_x - 1 >= 0 and last_y - 1 >= 0 and (last_x - 1, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x - 1, last_y - 1)])
+        if last_y - 1 >= 0 and (last_x, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x, last_y - 1)])
+        if  last_x + 1 < self.m and last_y - 1 >= 0 and (last_x + 1, last_y - 1) not in cells_set:
+            route_extended.append(route + [(last_x + 1, last_y - 1)])
+
+        return route_extended
+
+    def search_words(self):
+
+        '''Return all words found in the grid from given dictionary
+        Optimized by stopping searching longer route than maximum length of the dictionary word
+        '''
+
+        all_cells = [[(i, j)] for i in range(self.m) for j in range(self.n)]
+        routes = deque(all_cells)
+        words_found = set()
+        while routes:

```

Comment:
The choice of the approach depends on the size of the grid and the amount of words in the vocabulary indeed. So my solution is optimized for large vocabulary, but less for larger grids. Knowing the largest length of the word M helps to stop the traversal earlier in case larger grids though. If one word is super-long, that will decrease the performance. 

Alternative approach: On the other hand if the vocabulary is small, for each word we can explore several number routes of max len of this word M_i, that is much less than M_i choose (n*m) routes for word i, where grid is of n by m dimension.